sql: allow secondary tenants to set/show zone configurations

arulajmani opened the pull request:
Part of https://github.com/cockroachdb/cockroach/issues/67679. The zone configurations themselves have no effect right now. See individual commits for details.

arulajmani committed:
sql: add `system.zones` for secondary tenants

This patch bootstraps `system.zones` for secondary tenants and adds a
migration for existing tenants. We also seed secondary tenants with the
RANGE DEFAULT zone configuration.

Release note: None

arulajmani committed:
sql: enable secondary tenants to interact with zone configurations

This patch removes all checks that blocked secondary tenants from
writing to and reading from system.zones and replaces them with
checks to ensure `system.zones` exists. This allows us to run our
logic test suite for zone configurations in multi-tenant mode as well.

Currently, we don't validate node localities for secondary tenants when
setting zone configurations. A followup patch will address this.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/69169)
<!-- Reviewable:end -->

irfansharif reviewed (commented):
Just a bit unclear on how we're expecting to feature gate this. Everything else LGTM; deferring to Andrew's rubber stamp.

> Reviewed 11 of 11 files at r1, 27 of 27 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69169#-:-MhVLv7Y-ZvKKvGIkGS1:b-moe1fh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, and @RichardJCai)

___
*[pkg/migration/migrations/migrations.go, line 125 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69169#-MhVEqbx48uoLXyanOt-:-MhVEqbx48uoLXyanOt0:b-cuw07v) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4bdfe2cd58c7a7dfac35e3bd4178d420dd93de8/pkg/migration/migrations/migrations.go#L125)):*
> ```Go
> 	),
> 	migration.NewTenantMigration(
> 		"add system.zones table",
> ```

nit: add "for secondary tenants"
___
*[pkg/sql/tenant.go, line 160 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69169#-MhVFcpq2sTYzEcNzkWk:-MhVFcpq2sTYzEcNzkWl:b-qn5gpf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4bdfe2cd58c7a7dfac35e3bd4178d420dd93de8/pkg/sql/tenant.go#L160)):*
> ```Go
> 	schema := bootstrap.MakeMetadataSchema(
> 		codec,
> 		p.ExtendedEvalContext().ExecCfg.DefaultZoneConfig, /* defaultZoneConfig */
> ```

Curious to hear others' thoughts on what the tenant's range default should be. This seems fine but is static. If we want to be able to configure new tenant's RANGE DEFAULT, I thought it could be fun to overload RANGE TENANT for it. Worth a TODO perhaps?
___
*[pkg/sql/zone_config.go, line 425 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69169#-MhVIjr_DHIIUpPv5P6S:-MhVIjr_DHIIUpPv5P6T:brzei3a) ([raw file](https://github.com/cockroachdb/cockroach/blob/345d00c27a424762c14c9155f4cb94cac4099ccb/pkg/sql/zone_config.go#L425)):*
> ```Go
> // TODO(arul): All usages of this function can be removed in 22.1 as then all
> // all secondary tenants would have `system.zones`.
> func ZonesTableExists(ctx context.Context, codec keys.SQLCodec, settings *cluster.Settings) bool {
> ```

Though convenient, I think the signature hides away the fact that we're making a version check. Don't we also want to add a hidden feature flag that's disabled by default? We could do that check here too; just name this differently and adjust the surrounding comments at the callsites.
___
*[pkg/sql/catalog/bootstrap/metadata.go, line 285 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69169#-MhVH1x48t_AqQRHyHFo:-MhVH1x48t_AqQRHyHFp:bufucq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4bdfe2cd58c7a7dfac35e3bd4178d420dd93de8/pkg/sql/catalog/bootstrap/metadata.go#L285)):*
> ```Go
> // CreateZoneConfigKV creates a kv pair for the zone config for the given key
> // and config value.
> func CreateZoneConfigKV(
> ```

Does this need to be exported?
___
*[pkg/sql/catalog/bootstrap/metadata.go, line 310 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69169#-MhVHFYj2YTnRbYWtYrn:-MhVHFYj2YTnRbYWtYro:b-g4y34p) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4bdfe2cd58c7a7dfac35e3bd4178d420dd93de8/pkg/sql/catalog/bootstrap/metadata.go#L310)):*
> ```Go
> 		CreateZoneConfigKV(keys.RootNamespaceID, codec, defaultZoneConfig))
> 
> 	if codec.ForSystemTenant() {
> ```

Reduce a level of indentation for everything below by returning early if we're not the system tenant.
___
*[pkg/sql/gcjob/descriptor_utils.go, line 68 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69169#-MhVJMEbDX_VqP_rXubW:-MhVJMEbDX_VqP_rXubX:bq62vks) ([raw file](https://github.com/cockroachdb/cockroach/blob/345d00c27a424762c14c9155f4cb94cac4099ccb/pkg/sql/gcjob/descriptor_utils.go#L68)):*
> ```Go
> 		return nil
> 	}
> 	if !sql.ZonesTableExists(ctx, codec, settings) {
> ```

For everywhere we do this check, to feature gate this correctly I think we want some way to disable secondary tenants to muck about with zone configs unless a one-way hidden experimental flag/cluster setting was toggled. With this PR as is, tenants are able to write set zone configs which will confusingly never get acted on. I don't know if that's what we want.
___
*[pkg/sql/logictest/testdata/logic_test/zone_config, line 1 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69169#-MhVKdV_7XupRmzmsPr5:-MhVKdV_7XupRmzmsPr6:bzftu3a) ([raw file](https://github.com/cockroachdb/cockroach/blob/65457f9df618177bfd56096c9be0194db2ef5cf8/pkg/sql/logictest/testdata/logic_test/zone_config#L1)):*
> ```
> # LogicTest: !3node-tenant(49854)
> ```

Sweet!
___
*[pkg/sql/logictest/testdata/logic_test/zone_config_system_tenant, line 8 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69169#-MhVKp797LOY2q04HsaI:-MhVKp797LOY2q04HsaJ:b-qepwmq) ([raw file](https://github.com/cockroachdb/cockroach/blob/345d00c27a424762c14c9155f4cb94cac4099ccb/pkg/sql/logictest/testdata/logic_test/zone_config_system_tenant#L8)):*
> ```
> 
> # TODO(arul): move this back to `zone_config` once we validate cluster regions
> # for tenants.
> ```

Want to add cluster region validation to #67679 to not lose track?


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69169)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @arulajmani, @irfansharif, and @RichardJCai)

___
*[pkg/sql/zone_config.go, line 425 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69169#-MhVIjr_DHIIUpPv5P6S:-MhW55dQAMQRns7y5_tV:bjk3oha) ([raw file](https://github.com/cockroachdb/cockroach/blob/345d00c27a424762c14c9155f4cb94cac4099ccb/pkg/sql/zone_config.go#L425)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Though convenient, I think the signature hides away the fact that we're making a version check. Don't we also want to add a hidden feature flag that's disabled by default? We could do that check here too; just name this differently and adjust the surrounding comments at the callsites.
</blockquote></details>

Do we want to bundle the hidden feature flag as part of this method or do we want it to be a bit more targeted? Consider the case where a tenant has laid down some zone configs before flipping this cluster setting. We probably want to GC them and allow for introspection, right? 

I was thinking that the only place this cluster setting would be consulted is when setting new zone configs. All other call sites should really just care about if the zones table exists or not. Wdyt? 
___
*[pkg/sql/catalog/bootstrap/metadata.go, line 285 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69169#-MhVH1x48t_AqQRHyHFo:-MhW8NjJDTIJrWjQz33V:bbwzt0a) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4bdfe2cd58c7a7dfac35e3bd4178d420dd93de8/pkg/sql/catalog/bootstrap/metadata.go#L285)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Does this need to be exported?
</blockquote></details>

Not anymore, good catch! 
___
*[pkg/sql/gcjob/descriptor_utils.go, line 68 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69169#-MhVJMEbDX_VqP_rXubW:-MhW7k4N1vmbKvw3zkrS:bfvi3do) ([raw file](https://github.com/cockroachdb/cockroach/blob/345d00c27a424762c14c9155f4cb94cac4099ccb/pkg/sql/gcjob/descriptor_utils.go#L68)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

For everywhere we do this check, to feature gate this correctly I think we want some way to disable secondary tenants to muck about with zone configs unless a one-way hidden experimental flag/cluster setting was toggled. With this PR as is, tenants are able to write set zone configs which will confusingly never get acted on. I don't know if that's what we want.
</blockquote></details>

In this particular instance we probably want to delete the zone config regardless of the cluster setting though, right? Otherwise we'll leave orphaned entries in `system.zones`. See my comment above, I think we probably just want to gate new zone config writes behind that cluster setting instead of every place we do this check. 


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69169#-:-MhWlfWf3BubQWA7ywIb:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @arulajmani, @irfansharif, and @RichardJCai)

___
*[pkg/migration/migrations/zones.go, line 27 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69169#-MhWlU8h8wS_rif94fgl:-MhWlU8h8wS_rif94fgm:b-uocshz) ([raw file](https://github.com/cockroachdb/cockroach/blob/345d00c27a424762c14c9155f4cb94cac4099ccb/pkg/migration/migrations/zones.go#L27)):*
> ```Go
> // zonesTableForSecondaryTenants adds system.zones to secondary tenants and
> // seeds it.
> func zonesTableForSecondaryTenants(
> ```

if I were being really nitpicky I'd ask for a test of this 
___
*[pkg/sql/tenant.go, line 160 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69169#-MhVFcpq2sTYzEcNzkWk:-MhWi4874pysNLIoVSvB:bczeyhx) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4bdfe2cd58c7a7dfac35e3bd4178d420dd93de8/pkg/sql/tenant.go#L160)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Curious to hear others' thoughts on what the tenant's range default should be. This seems fine but is static. If we want to be able to configure new tenant's RANGE DEFAULT, I thought it could be fun to overload RANGE TENANT for it. Worth a TODO perhaps?
</blockquote></details>

I think configuring this warrants an issue in github. 
___
*[pkg/sql/zone_config.go, line 425 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69169#-MhVIjr_DHIIUpPv5P6S:-MhWinyX5CkM_KErWj63:b-onkvrl) ([raw file](https://github.com/cockroachdb/cockroach/blob/345d00c27a424762c14c9155f4cb94cac4099ccb/pkg/sql/zone_config.go#L425)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Do we want to bundle the hidden feature flag as part of this method or do we want it to be a bit more targeted? Consider the case where a tenant has laid down some zone configs before flipping this cluster setting. We probably want to GC them and allow for introspection, right? 

I was thinking that the only place this cluster setting would be consulted is when setting new zone configs. All other call sites should really just care about if the zones table exists or not. Wdyt? 
</blockquote></details>

You can just pass in the `clusterversion.Handle` if you wanted. 

I don't feel strongly on whether this is the spot that also reads a testing knob. I haven't gotten to such a knob yet or where this is called. 
___
*[pkg/sql/gcjob/descriptor_utils.go, line 68 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69169#-MhVJMEbDX_VqP_rXubW:-MhWjXdTBg6wGPUR90kD:b-n7o2um) ([raw file](https://github.com/cockroachdb/cockroach/blob/345d00c27a424762c14c9155f4cb94cac4099ccb/pkg/sql/gcjob/descriptor_utils.go#L68)):*
> I think we probably just want to gate new zone config writes behind that cluster setting instead of every place we do this check.
:thumbsup: 


<!-- Sent from Reviewable.io -->

arulajmani committed:
sql: gate setting zone configs behind a cluster setting for tenants

This patch gates setting zone configurations behind a the
`sql.zone_configs.allow_for_secondary_tenant.enabled` cluster setting
for secondary tenants. For now, this thing is not public.

Release note: None

arulajmani reviewed (commented):
Added a third commit that adds a cluster setting to gate secondary tenants from writing to `system.zones`. It's off by default, has no effect for the system tenant, and does not affect already existing zone configs -- only new ones. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/69169#-:-Mh_Ai_Q0VIZYrtLi1Tc:bcxbzqs)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @arulajmani, @irfansharif, and @RichardJCai)

___
*[pkg/sql/zone_config.go, line 425 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69169#-MhVIjr_DHIIUpPv5P6S:-Mh_AWEb-oov39mSeU9p:b-hfuyuj) ([raw file](https://github.com/cockroachdb/cockroach/blob/345d00c27a424762c14c9155f4cb94cac4099ccb/pkg/sql/zone_config.go#L425)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

You can just pass in the `clusterversion.Handle` if you wanted. 

I don't feel strongly on whether this is the spot that also reads a testing knob. I haven't gotten to such a knob yet or where this is called. 
</blockquote></details>

Done.

For completeness, it isn't a testing knob but a cluster setting. Wasn't there when you reviewed this, but I've tacked on a commit here that adds the cluster setting and checks for it when setting zone configurations for secondary tenants. 
___
*[pkg/sql/catalog/bootstrap/metadata.go, line 310 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/69169#-MhVHFYj2YTnRbYWtYrn:-MhZmVJD0drRIWBQALnZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4bdfe2cd58c7a7dfac35e3bd4178d420dd93de8/pkg/sql/catalog/bootstrap/metadata.go#L310)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Reduce a level of indentation for everything below by returning early if we're not the system tenant.
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/zone_config_system_tenant, line 8 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/69169#-MhVKp797LOY2q04HsaI:-MhZoIjE9t17kxSq7BH_:b-sb12va) ([raw file](https://github.com/cockroachdb/cockroach/blob/345d00c27a424762c14c9155f4cb94cac4099ccb/pkg/sql/logictest/testdata/logic_test/zone_config_system_tenant#L8)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Want to add cluster region validation to #67679 to not lose track?
</blockquote></details>

Filed: https://github.com/cockroachdb/cockroach/issues/69199 and added it to the tracking issue. 


<!-- Sent from Reviewable.io -->

irfansharif commented on pkg/spanconfig/spanconfigmanager/manager.go:
```diff
@@ -7,7 +7,6 @@
 // the Business Source License, use of this software will be governed
 // by the Apache License, Version 2.0, included in the file
 // licenses/APL.txt.
-
```
Accidental newline removal?

irfansharif commented on pkg/sql/set_zone_config.go:
Since this is an experimental cluster setting we want to hide from users, it's fine to not refer to the setting name. "zone configs disabled for multi-tenancy".

arulajmani replied:
Good point, I'll just re-use the "unsupported with multi-tenancy" here.

irfansharif commented on pkg/sql/zone_config.go:
```diff
@@ -41,6 +42,18 @@ func init() {
 
 var errNoZoneConfigApplies = errors.New("no zone config applies")
 
+const secondaryTenantsZoneConfigsEnabledSettingName = "sql.zone_configs.allow_for_secondary_tenant.enabled"
```
Need to call out the experimental nature, how about "sql.defaults.experimental_multitenant_zone_configs.enabled" instead? And putting it next to the other experimental sql cluster settings over at pkg/sql/exec_util.go?

arulajmani replied:
Talked offline, changing it to `sql.zone_configs.experimental_allow_secondary_tenants.enabled`

arulajmani commented:
TFTRs! 

bors r=irfansharif,ajwerner
